lib/qos/qos.c

```diff
@@ -0,0 +1,335 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr.h>
+#include <sys/slist.h>
+
+#include "qos.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(qos, CONFIG_QOS_LOG_LEVEL);
+
+#define INVALID -1
+
+/* Lookup table for QoS backoff messages. */
+uint32_t backoff_delay_lookup_sec[] = {
+	16, 32, 64
+};
+
+/* Structure used to keep track of pending messages. Used in combination with a linked list. */
+struct qos_metadata {
+	/* Mandatory variable used to construct a linked list. */
+	sys_snode_t header;
+
+	/* Message associated with each node in the linked list. */
+	struct qos_data message;
+};
+
+/* Structure containing internal variables in the library.  */
+static struct ctx {
+	/* Library event handler. Used in callbacks to the caller. */
+	qos_evt_handler_t app_evt_handler;
+
+	/* Number of times the message retransmission has been invoked by the library.
+	 * Used to index backoff delay.
+	 */
+	uint8_t retry_count;
+
+	/* Internal array of pending messages. Used in combination with linked list. */
+	struct qos_metadata list_internal[CONFIG_QOS_PENDING_MESSAGES_MAX];
+
+	/* Linked list used to keep track of pending messages. */
+	sys_slist_t pending_list;
+
+	/* Variable used to prevent multiple library initializations. */
+	bool initialized;
+
+	/* Delayed work usd to schedule a backoff timer. */
+	struct k_work_delayable timeout_handler_work;
+} ctx;
+
+/* Mutex used to protect internal library variables. */
+static K_MUTEX_DEFINE(ctx_lock);
```

Comment:
The mutex is used in some functions, but I think there are several places which should also be protected with this mutex.

Reply:
I think most of the public APIs should have this lock. And the `timeout_handler_work_fn` also actually. Correct me if I'm wrong.

Reply:
Basically it should be used whenever accessing something in `ctx`, but I think protecting access to the lists might be enough.

Reply:
Updated mutex handling

Reply:
Is it possible that the application calls for example `qos_message_remove()` from the application event handler? Just asking because when the event handler is called, the mutex is always locked. If another API function would be called from the event handler, it would try to lock the mutex again, resulting in a deadlock, I think.

Reply:
That's true...😱 There might be use cases where an application might want to do this.
We could just do an unlock before the callback and then do a lock again once the callback returns. I think that should work. I need to test the scenario where a function call in the callback invokes another callback. I'm not sure what happens then.

Reply:
Yes, this is tricky because usually when the callback is called the code is iterating through the list and the callback might do something which causes the list to be modified. I hope it works. 😄

Reply:
Did some digging and it seems that this should be no issue as long as its the same thread that locks/unlocks the mutex subsequently. Calling a QoS lib API in the callback shouldn't be a problem based on this. Testing and API docs for the lock and unlock call also suggest this.

k_mutex_unlock:
`Unlock a mutex. This routine unlocks mutex. The mutex must already be locked by the calling thread. The mutex cannot be claimed by another thread until it has been unlocked by the calling thread as many times as it was previously locked by that thread. `

Reply:
Great, I didn't realize that the mutex can be locked multiple times by the same thread. This should be fine then, as long as modifying the list while iterating it works properly.